File: scripts/introspect-missing-tables.ts
#!/usr/bin/env tsx
/**
 * Introspect missing tables from Neon database to reconstruct schema definitions
 */
import { drizzle } from 'drizzle-orm/postgres-js';
import postgres from 'postgres';

const DATABASE_URL = process.env.DATABASE_URL;
if (!DATABASE_URL) {
  throw new Error('DATABASE_URL not set');
}

const client = postgres(DATABASE_URL);
const db = drizzle(client);

async function introspectTable(tableName: string) {
  console.log(`\n=== ${tableName} ===`);
  
  // Get column definitions
  const columns = await client`
    SELECT 
      column_name,
      data_type,
      character_maximum_length,
      is_nullable,
      column_default
    FROM information_schema.columns
    WHERE table_schema = 'public' AND table_name = ${tableName}
    ORDER BY ordinal_position;
  `;
  
  console.log('Columns:');
  columns.forEach(col => {
    const nullable = col.is_nullable === 'YES' ? '' : 'NOT NULL';
    const length = col.character_maximum_length ? `(${col.character_maximum_length})` : '';
    const defaultVal = col.column_default ? `DEFAULT ${col.column_default}` : '';
    console.log(`  ${col.column_name}: ${col.data_type}${length} ${nullable} ${defaultVal}`.trim());
  });
  
  // Get indexes
  const indexes = await client`
    SELECT 
      indexname,
      indexdef
    FROM pg_indexes
    WHERE schemaname = 'public' AND tablename = ${tableName};
  `;
  
  if (indexes.length > 0) {
    console.log('Indexes:');
    indexes.forEach(idx => console.log(`  ${idx.indexname}: ${idx.indexdef}`));
  }
  
  // Get row count
  const count = await client`SELECT COUNT(*) as count FROM ${client(tableName)}`;
  console.log(`Row count: ${count[0].count}`);
}

async function main() {
  const missingTables = [
    'qig_metadata',
    'governance_proposals', 
    'tool_requests',
    'pattern_discoveries',
    'vocabulary_stats',
    'federation_peers',
    'passphrase_vocabulary'
  ];
  
  for (const table of missingTables) {
    try {
      await introspectTable(table);
    } catch (err) {
      console.error(`Error introspecting ${table}:`, err);
    }
  }
  
  await client.end();
}

main().catch(console.error);

File: scripts/introspect_missing_tables.py
#!/usr/bin/env python3
"""
Introspect missing tables from Neon database to reconstruct schema definitions
"""
import os
import psycopg2
from psycopg2.extras import RealDictCursor

DATABASE_URL = os.environ.get('DATABASE_URL')
if not DATABASE_URL:
    raise ValueError('DATABASE_URL not set')

def introspect_table(cursor, table_name):
    print(f"\n=== {table_name} ===")
    
    # Get column definitions
    cursor.execute("""
        SELECT 
            column_name,
            data_type,
            character_maximum_length,
            is_nullable,
            column_default
        FROM information_schema.columns
        WHERE table_schema = 'public' AND table_name = %s
        ORDER BY ordinal_position;
    """, (table_name,))
    
    columns = cursor.fetchall()
    if not columns:
        print(f"  Table {table_name} not found")
        return
    
    print("Columns:")
    for col in columns:
        nullable = '' if col['is_nullable'] == 'YES' else 'NOT NULL'
        length = f"({col['character_maximum_length']})" if col['character_maximum_length'] else ''
        default = f"DEFAULT {col['column_default']}" if col['column_default'] else ''
        print(f"  {col['column_name']}: {col['data_type']}{length} {nullable} {default}".strip())
    
    # Get indexes
    cursor.execute("""
        SELECT indexname, indexdef
        FROM pg_indexes
        WHERE schemaname = 'public' AND tablename = %s;
    """, (table_name,))
    
    indexes = cursor.fetchall()
    if indexes:
        print("Indexes:")
        for idx in indexes:
            print(f"  {idx['indexname']}: {idx['indexdef']}")
    
    # Get row count
    cursor.execute(f"SELECT COUNT(*) as count FROM {table_name}")
    count = cursor.fetchone()
    print(f"Row count: {count['count']}")

def main():
    missing_tables = [
        'qig_metadata',
        'governance_proposals',
        'tool_requests',
        'pattern_discoveries',
        'vocabulary_stats',
        'federation_peers',
        'passphrase_vocabulary'
    ]
    
    conn = psycopg2.connect(DATABASE_URL)
    cursor = conn.cursor(cursor_factory=RealDictCursor)
    
    try:
        for table in missing_tables:
            try:
                introspect_table(cursor, table)
            except Exception as e:
                print(f"Error introspecting {table}: {e}")
    finally:
        cursor.close()
        conn.close()

if __name__ == '__main__':
    main()
